fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/client/item/v1/ArmorTextureRegistry.java

```diff
@@ -0,0 +1,100 @@
+import net.minecraft.item.ItemStack;
+
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.fabricmc.fabric.impl.client.item.ArmorTextureRegistryImpl;
+
+/**
+ * A class for registering custom armor textures for {@link ItemConvertible}, to be provided by a {@link Provider}.
+ *
+ * <p>
+ * This can be used to replace existing vanilla armor textures conditionally, however each {@link net.minecraft.item.Item}
+ * instance can only allow one {@link Provider}.
+ * </p>
+ *
+ * <p>
+ * Since armor textures identifier in vanilla is hardcoded to be in the {@code minecraft} namespace, this registry can also be
+ * used to use a custom namespace if desired.
+ * </p>
+ */
+@Environment(EnvType.CLIENT)
+public final class ArmorTextureRegistry {
+	private ArmorTextureRegistry() {
+	}
+
+	/**
+	 * Registers a provider for custom texture models for an item.
+	 *
+	 * @param provider the provider for the texture
+	 * @param items    the items to be registered for
+	 */
+	public static void register(/* @Nullable */ Provider provider, ItemConvertible... items) {
+		register(provider, Arrays.asList(items));
+	}
+
+	/**
+	 * Registers a provider for custom texture models for an item.
+	 *
+	 * @param provider the provider for the texture
+	 * @param items    the items to be registered for
+	 */
+	public static void register(/* @Nullable */ Provider provider, Iterable<ItemConvertible> items) {
+		ArmorTextureRegistryImpl.register(provider, Objects.requireNonNull(items));
+	}
+
+	/**
+	 * Gets the armor texture identifier in string, to be converted to {@link net.minecraft.util.Identifier}.
+	 *
+	 * @param entity      The entity equipping the armor
+	 * @param stack       The item stack of the armor
+	 * @param slot        The slot which the armor is in
+	 * @param secondLayer Whether the texture should use the second layer, only true when the {@code slot} is {@link EquipmentSlot#LEGS}
+	 * @param suffix      The nullable suffix of the texture, may be "overlay" in vanilla when the item is {@link net.minecraft.item.DyeableArmorItem}
+	 * @return the custom armor texture identifier, return null to use the vanilla ones. Defaulted to the item's registry id.
+	 */
+	/* @Nullable */
+	public static String getArmorTexture(LivingEntity entity, ItemStack stack, EquipmentSlot slot, boolean secondLayer, /* @Nullable */  String suffix) {
+		return ArmorTextureRegistryImpl.getArmorTexture(entity, stack, slot, secondLayer, suffix);
+	}
+
+	@FunctionalInterface
+	@Environment(EnvType.CLIENT)
+	public interface Provider {
+		/**
+		 * Gets the armor texture identifier in string, to be converted to {@link net.minecraft.util.Identifier}.
+		 *
+		 * @param entity      The entity equipping the armor
+		 * @param stack       The item stack of the armor
+		 * @param slot        The slot which the armor is in
+		 * @param secondLayer Whether the texture should use the second layer, only true when the {@code slot} is {@link EquipmentSlot#LEGS}
```

Comment:
I don't think we should pass secondLayer like that while also supplying the slot. It is just a weird artefact of MC's implementation. Likewise suffix should probably be replaced with boolean overlay to denote the 2nd render pass without dye.

We could instead supply the full vanilla path in the `textures/models/armor/[..]_layer_[..].png` form to make this easy for namespace replacement use cases, the proposed ArmorModelRegistry already does something similar with defaultModel.